add model inversion post (#83)
rstudio/ai-blog: _posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd

---
title: "Hacking deep learning: model inversion attack by example"
description: > 

>   Compared to other applications, deep learning models might not seem too likely as victims of privacy attacks. However, methods exist to determine whether an entity was used in the training set (an adversarial attack called member inference), and techniques subsumed under "model inversion" allow to spy on raw data input given just model output (and sometimes, context information). This post shows an end-to-end example of model inversion, and explores mitigation strategies using TensorFlow Privacy. 

Reviewer: > allow to spy on raw data input given just model output (and sometimes, context information)

I think you need a word between "allow" and "to spy" (allow _what_ to spy?)

Author: hm... or another verb...? reconstruct, I think...

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydanamodelinversion
bibliography: bibliography.bib
date: 05-15-2020
categories:
  - R
  - Privacy & Security
  - TensorFlow/Keras
output:
  distill::distill_article:
    self_contained: false
preview: images/results.png
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

How private are individual data in the context of machine learning models? The data used to train the model, say. There are

> types of models where the answer is simple. Take k-nearest-neighbors, for example. There *is* not even a model without the

Reviewer: ```suggestion
types of models where the answer is simple. Take k-nearest-neighbors, for example. There *is not* even a model without the
```

complete dataset. Or support vector machines. There is no model without the support vectors. But neural networks? They're just
some composition of functions, -- no data included.

> Same with data fed to a deployed deep learning model. It's pretty unlikely one could invert the final softmax output from a

Reviewer: ```suggestion
The same is true for data fed to a deployed deep-learning model. It's pretty unlikely one could invert the final softmax output from a
```

big ResNet and get back the raw input data.

In theory, then, "hacking" a standard neural net to spy on input data sounds illusory. In practice, however, there is always
some real-world *context*. The context may be other datasets, publicly available, that can be linked to the "private" data in
question. This is a popular showcase used in advocating for differential privacy[@Dwork2006]: Take an "anonymized" dataset,
dig up complementary information from public sources, and de-anonymize records ad libitum. Some context in that sense will
often be used in "black-box" attacks, ones that presuppose no insider information about the model to be hacked.

But context can also be structural, such as in the scenario demonstrated in this post. For example, assume a distributed

> model, where sets lof layers run on different devices (embedded devices or mobile phones, for example). (A scenario like that

Reviewer: ```suggestion
model, where sets of layers run on different devices (embedded devices or mobile phones, for example). (A scenario like that
```

Reviewer: Having a parenthetical right after a sentence that ends with a parenthetical just feels off to me.

is sometimes seen as "white-box"[@7536387], but in common understanding, white-box attacks probably presuppose some more
insider knowledge, such as access to model architecture or even, weights. I'd therefore prefer calling this white-ish at
most.) --- Now assume that in this context, it is possible to intercept, and interact with, a system that executes the deeper
layers of the model. Based on that system's intermediate-level output, it is possible to perform *model inversion*[@Fred],

> that is, reconstruct the input data fed into the system.

Reviewer: ```suggestion
that is, to reconstruct the input data fed into the system.
```

> In this post, we'll demonstrate such a model inversion attack, basically porting the approach given in a

Reviewer: ```suggestion
In this post, we'll demonstrate such a model-inversion attack, basically porting the approach given in a
```

[notebook](https://github.com/OpenMined/PySyft/blob/master/examples/tutorials/advanced/privacy_attacks/Tutorial%201%20-%20Black%20box%20model%20inversion.ipynb Sy)
found in the [PySyft](https://github.com/OpenMined/PySyft) repository. We then experiment with different levels of

> $\epsilon$-privacy, exploring impact on reconstruction success. This second part will make use of TensorFlow Privacy,

Reviewer: Add link to TensorFlow Privacy here for first mention?

introduced in a [previous blog post](https://blogs.rstudio.com/ai/posts/2019-12-20-differential-privacy/).

## Part 1: Model inversion in action

### Example dataset: All the world's letters[^1]

[^1]: Don't take *all* literally please; it's just a nice phrase.

The overall process of model inversion used here is the following. With no, or scarcely any, insider knowledge about a model,
-- but given opportunities to repeatedly query it --, I want to learn how to reconstruct unknown inputs based on just model

> outputs . Independently of original model training, this, too, is a training process; however, in general it will not involve

Reviewer: ```suggestion
outputs . Independent of original model training, this, too, is a training process; however, in general it will not involve
```

the original data, as those won't be publicly available. Still, for best success, the attacker model is trained with data as
similar as possible to the original training data assumed. Thinking of images, for example, and presupposing the popular view

> of successive layers representing successively coarse-grained features, we want that the surrogate data to share as many

Reviewer: ```suggestion
of successive layers representing successively coarse-grained features, we want the surrogate data to share as many
```

representation spaces with the real data as possible -- up to the very highest layers before final classification, ideally.

If we wanted to use classical MNIST as an example, one thing we could do is

> to only use some of the digits for training the "real" model; and the rest, for training the adversary. Let's try something

Reviewer: ```suggestion
to use only some of the digits for training the "real" model; and the rest, for training the adversary. Let's try something
```

different though, something that might make the undertaking harder as well as easier at the same time. Harder, because the

> dataset features exemplars more complex than MNIST digits; easier because of the same reason: More could possibly be learned,
> by the adversary, from a complex task.

Reviewer: ```suggestion
dataset features exemplars more complex than MNIST digits; easier because of the same reason: More could possibly be learned
by the adversary from a complex task.
```

Originally designed to develop a machine model of concept learning and generalization [@Lake1332], the
[OmniGlot](https://github.com/brendenlake/omniglot/) dataset incorporates characters from fifty alphabets, split into two
disjoint groups of thirty and twenty alphabets each. We'll use the group of twenty to train our target model. Here is a
sample:

```{r, eval=TRUE, echo=FALSE, fig.cap = "Sample from the twenty-alphabet set used to train the target model (originally: 'evaluation set')"}
knitr::include_graphics("images/training_set_images.png")
```

The group of thirty we don't use; instead, we'll employ two small five-alphabet collections to train the adversary and to test
reconstruction, respectively. (These small subsets of the original "big" thirty-alphabet set are again disjoint.)

> Here first is a sample from the set used to train the adversary.

Reviewer: ```suggestion
Here, first, is a sample from the set used to train the adversary.
```

knitr::include_graphics("images/actual_test_images.png")
```

Conveniently, we can use [tfds](https://github.com/rstudio/tfds), the R wrapper to TensorFlow Datasets, to load those subsets:

```{r}
library(reticulate)
library(keras)
library(tfdatasets)
library(tfautograph)
library(tfds)

library(purrr)

# we'll use this to train the target model
# n = 13180
omni_train <- tfds$load("omniglot", split = "test")

# this is used to train the adversary
# n = 2720
omni_spy <- tfds$load("omniglot", split = "small1")

# this we'll use for testing
# n = 3120
omni_test <- tfds$load("omniglot", split = "small2")

```

Now first, we train the target model.

### Train target model

The dataset originally has four columns: the image, of size 105 x 105; an alphabet id and a within-dataset character id; and a
label. For our use case, we're not really interested in the task the target model was/is used for; we just want to get at the
data. Basically, whatever task we choose, it is not much more than a dummy task. So, let's just say we train the target to
classify characters *by alphabet*.

> We thus throw out all unneeded features, keeping just the alphabet id and the image itself:

Reviewer: ```suggestion
Thus, we throw out all unneeded features, keeping just the alphabet id and the image itself:
```

preprocess_image <- function(image) {
  image %>%
    tf$cast(dtype = tf$float32) %>%
    tf$truediv(y = 255) %>%
    tf$image$rgb_to_grayscale()
}

# use the first 11000 images for training
train_ds <- omni_train %>% 
  dataset_take(11000) %>%
  dataset_map(function(record) {
    record$image <- preprocess_image(record$image)
    list(record$image, record$alphabet)}) %>%
  dataset_shuffle(1000) %>% 
  dataset_batch(32)

# use the remaining 2180 records for validation
val_ds <- omni_train %>% 
  dataset_skip(11000) %>%
  dataset_map(function(record) {
    record$image <- preprocess_image(record$image)
    list(record$image, record$alphabet)}) %>%
  dataset_batch(32)
```

The model consists of two parts. The first is imagined to run in a distributed fashion; for example, on mobile devices (stage
one). These devices then send model outputs to a central server, where final results are computed (stage two). Sure, you will
be thinking, this is a convenient setup for our scenario: If we intercept stage one results, we -- most probably -- gain
access to richer information than what is contained in a model's final output layer. --- That is correct, but the scenario is

> less contrived than one might assume. Just like federated learning [@McMahanMRA16], it fulfills important desiderata: Actual

Reviewer: I love the word desiderata (I literally have a GitHub repo by that name), but I think it's a bit esoteric in this context. 🤓

training data never leaves the devices, thus staying (in theory!) private; at the same time, ingoing traffic to the server is
significantly reduced.

In our example setup, the on-device model is a convnet, while the server model is a simple feedforward network.

We link both together as a *TargetModel* that when called normally, will run both steps in succession. However, we'll be able

> to call `target_model$mobile_step()` separately, thus intercepting intermediate results.

Reviewer: ```suggestion
to call `target_model$mobile_step()` separately, thereby intercepting intermediate results.
```

Reviewer: Just mixing the words up a bit for kicks.

Author: right ... :-) i should consult the synonyms site more often ...

  layer_dropout(0.2) %>%
  layer_conv_2d(filters = 32, kernel_size = c(3, 3), activation = "relu") %>%
  layer_batch_normalization() %>%
  layer_max_pooling_2d(pool_size = c(2, 2), strides = 2) %>%
  layer_dropout(0.2) 

server_model <- keras_model_sequential() %>%
  layer_dense(units = 256, activation = "relu") %>%
  layer_flatten() %>%
  layer_dropout(0.2) %>% 
  # we have just 20 different ids, but they are not in lexicographic order
  layer_dense(units = 50, activation = "softmax")

target_model <- function() {
  keras_model_custom(name = "TargetModel", function(self) {
    
    self$on_device_model <-on_device_model
    self$server_model <- server_model
    self$mobile_step <- function(inputs) 
      self$on_device_model(inputs)
    self$server_step <- function(inputs)
      self$server_model(inputs)

    function(inputs, mask = NULL) {
      inputs %>% 
        self$mobile_step() %>%
        self$server_step()
    }
  })
  
}

model <- target_model()

```

The overall model is a Keras custom model, so we train it [TensorFlow 2.x -
style](https://tensorflow.rstudio.com/tutorials/advanced/). After ten epochs, training and validation accuracy are at

> ~0.84\ and~ 0.73, respectively -- not bad at all for a 20-class discrimination task.

Reviewer 2: This seems to be causing some weird rendering, intended?

Author: right, thanks! Will correct...

           "    Validation Accuracy", val_accuracy$result())
  
  train_loss$reset_states()
  train_accuracy$reset_states()
  val_loss$reset_states()
  val_accuracy$reset_states()
}))

for (epoch in 1:10) {
  cat("Epoch: ", epoch, " -----------\n")
  training_loop(train_ds, val_ds)  
}
```

    Epoch:  1  -----------
    Train accuracy 0.195090905     Validation Accuracy 0.376605511
    Epoch:  2  -----------
    Train accuracy 0.472272724     Validation Accuracy 0.5243119
    ...
    ...
    Epoch:  9  -----------
    Train accuracy 0.821454525     Validation Accuracy 0.720183492
    Epoch:  10  -----------
    Train accuracy 0.840454519     Validation Accuracy 0.726605475

Now, we train the adversary.

### Train adversary

The adversary's general strategy will be:

-   Feed its small, surrogate dataset to the on-device model. The output received can be regarded as a (highly) *compressed*
    version of the original images.
-   Pass that "compressed" version as input to its own model, which tries to reconstruct the original images from the sparse
    code.
-   Compare original images (those from the surrogate dataset) to the reconstruction pixel-wise. The goal is to minimize the

>     mean (squared, say) error.

Reviewer: I would embolden the action sentence in each of these, e.g., **Feed its small, surrogate dataset to the on-device model.**, or something to that effect, to make the list a bit more skimmable.

Author: makes sense!

training. Luckily, the workaround will be easy.

First, load (and possibly, install) libraries, taking care to disable TensorFlow V2 behavior.

```{r}
library(keras)
library(tensorflow)
# still necessary when working with TensorFlow Privacy, as of this writing
tf$compat$v1$disable_v2_behavior()

# if you don't have it installed:
# reticulate::py_install("tensorflow_privacy")
tfp <- import("tensorflow_privacy")

library(tfdatasets)
library(tfds)

library(purrr)
```

The training set is loaded, preprocessed and batched (nearly) as before.

```{r}
omni_train <- tfds$load("omniglot", split = "test")

batch_size <- 32

train_ds <- omni_train %>%
  dataset_take(11000) %>%
  dataset_map(function(record) {
    record$image <- preprocess_image(record$image)
    list(record$image, record$alphabet)}) %>%
  dataset_shuffle(1000) %>%
  # need dataset_repeat() when not eager
  dataset_repeat() %>%
  dataset_batch(batch_size)
```

### Train target model -- with TensorFlow Privacy

To train the target, we put the layers from both stages -- "mobile" and "server" -- into one sequential model. Note how we

> remove the dropout. This is because anyway, noise will be added during optimization.

Reviewer: ```suggestion
remove the dropout. This is because noise will be added during optimization anyway.
```

Author: hehe, rhetorical "syntax inversion" again ;-)

the target model is the same as before:

```{r}
on_device_model <- keras_model_sequential() %>%
  [...]

server_model <- keras_model_sequential() %>%
  [...]

target_model <- function() {
  keras_model_custom(name = "TargetModel", function(self) {
    
    self$on_device_model <-on_device_model
    self$server_model <- server_model
    self$mobile_step <- function(inputs) 
      self$on_device_model(inputs)
    self$server_step <- function(inputs)
      self$server_model(inputs)
    
    function(inputs, mask = NULL) {
      inputs %>% 
        self$mobile_step() %>%
        self$server_step()
    }
  })
}

intercepted <- target_model()

```

But now, we load the trained target's weights into the freshly defined model's "mobile stage":

```{r}

intercepted$on_device_model$load_weights("intercepted.hdf5")

```

And now, we're back to the old training routine. Testing setup is the same as before, as well.

So how well does the adversary perform with differential privacy added to the picture?

### Test adversary (against differentially private target)

Here, ordered by decreasing $\epsilon$, are the reconstructions. Again, we refrain from judging the results, for the same

> reasons as before: In real-world applications, whether privacy is preserved well "enough" will depend on the context.

Reviewer: ```suggestion
reasons as before: In real-world applications, whether privacy is preserved "well enough" will depend on the context.
```

knitr::include_graphics("images/intercepted_eps_12.5.png")
```

And the highest-$\epsilon$ one:

```{r, eval=TRUE, echo=FALSE, layout="l-body", fig.cap = "Reconstruction attempts from a setup where the target model was trained with an epsilon of 4.0."}
knitr::include_graphics("images/intercepted_eps_4.0.png")
```

## Conclusion

Throughout this post, we've refrained from "over-commenting" on results, and focused on the why-and-how instead. This is
because in an artificial setup, chosen to facilitate exposition of concepts and methods, there really is no objective frame of
reference. What is a good reconstruction? What is a good $\epsilon$? What constitutes a data breach? No-one knows.

In the real world, there is a context to everything -- there are people involved, the people whose data we're talking about.
There are organizations, regulations, laws. There are abstract principles, and there are implementations; different
implementations of the same "idea" can differ.

As in machine learning overall, research papers on privacy-, ethics- or otherwise society-related topics are full of LaTeX

> formulae. Between the math, let's not forget the people.

Reviewer: ```suggestion
formulae. Amid the math, let's not forget the people.
```

Author: oooh right...
